### PR TITLE
Convert between vsphere datastore format to URLs and back

### PR DESCRIPTION
**Pull request by fdawg4l**
Added helpers to convert between the two formats.

### CODE REVIEW

File: lib/portlayer/storage/vsphere/datastore.go [hunk 1]
```diff
@@ -15,9 +15,11 @@
 package vsphere
 
 import (
+	"fmt"
 	"io"
 	"net/url"
 	"path"
+	"regexp"
```
Comment (caglar10ur) [inline lib/portlayer/storage/vsphere/datastore.go]: https://imgs.xkcd.com/comics/perl_problems.png :)

**Comment by caglar10ur** [comment]
LGTM

File: lib/portlayer/storage/vsphere/datastore.go (hunk 1 above)
Comment (fdawg4l) [inline lib/portlayer/storage/vsphere/datastore.go]: @mavery helped me with the expression. Also, we know a Perl celebrity. True story.

File: lib/portlayer/storage/vsphere/datastore.go [hunk 2]
```diff
@@ -301,3 +303,36 @@ func (d *Datastore) mkRootDir(ctx context.Context, rootdir string) error {
...
+
+// Converts `[datastore] /path` to URL
+func DatastoreToURL(ds string) (*url.URL, error) {
+	matches := datastoreFormat.FindAllStringSubmatch(ds, -1)
+
+	if len(matches) != 1 || len(matches[0]) != 3 {
+		return nil, fmt.Errorf("invalid datastore format: %s", ds)
+	}
+
+	datastoreName := matches[0][1]
```
Comment (hmahmood) [inline lib/portlayer/storage/vsphere/datastore.go]: Is empty datastore name allowed? Your regexp allows that ('*' vs. '+').

Comment (hmahmood) [inline lib/portlayer/storage/vsphere/datastore.go]: Also, spaces allowed in datastore path? Can you omit it altogether? Your regex allows those cases too.

File: lib/portlayer/storage/vsphere/datastore.go [hunk 3]
```diff
@@ -301,3 +303,36 @@ func (d *Datastore) mkRootDir(ctx context.Context, rootdir string) error {
 func (d *Datastore) rootDir() string {
 	return strings.SplitN(d.RootURL, " ", 2)[1]
 }
+
+// Parse the datastore format ([datastore1] /path/to/thing) to groups.
+var datastoreFormat = regexp.MustCompile(`\[([\w\d]*)\]\ (.*)?`)
```
Comment (hickeng) [inline lib/portlayer/storage/vsphere/datastore.go]: Is this deliberately allowing `[] /path/to/thing` and `[]`?

`(.*)?` - ? is redundant as .* will greedily consume everything up to the end of the line so it's not possible to have more than one group here.

File: lib/portlayer/storage/vsphere/datastore_test.go [hunk 4]
```diff
@@ -127,6 +127,36 @@ func TestDatastoreMkdirAndLs(t *testing.T) {
...
 }
 
+func TestDatastoreToURLParsing(t *testing.T) {
+	expected := "ds://datastore1/path/to/thing"
+
+	input := []string{
+		"[datastore1] /path/to/thing",
+		"[datastore1] path/to/thing",
+		"[datastore1] ///path////to/thing",
+	}
```
Comment (hickeng) [inline lib/portlayer/storage/vsphere/datastore_test.go]: Add the following at a minimum if intended by expression:
```
"[]",
"[] /path/to/thing",

File: lib/portlayer/storage/vsphere/datastore.go (hunk 3 above)
Comment (fdawg4l) [inline lib/portlayer/storage/vsphere/datastore.go]: You're regex is stronger than mine.  Lets chat offline and you can help me craft something smarter.

**Comment by fdawg4l** [comment]
Going to drop this and allow @gigawhitlocks to grab this, fix it, and use it in his vic-machine changes since it'd be really useful there and i'm out of cycles to drive this to completion.  Thanks Ian!

**Pull request closed without merging**